components/scream/src/share/scream_logging.hpp

```diff
@@ -0,0 +1,271 @@
+#include <string>
+#include <vector>
+#include <stdexcept>
+#include "scream_config.h"
+
+#if LOG_STACK_TRACE
+#include <execinfo.h>
+#endif
+
+namespace scream {
+namespace util {
+struct Error : public std::runtime_error {
+  explicit Error(const std::string &s) : std::runtime_error(s) {}
+};
+}  // namespace util
+}  // namespace scream
+
+#include <assert.h>
+#include <iostream>
+#include <sstream>
+#include <ctime>
+
+namespace scream {
+namespace util   {
+
+inline void InitLogging(const char*) {
+  // DO NOTHING
+}
+
+class LogCheckError {
+ public:
+  LogCheckError() : str(nullptr) {}
+  explicit LogCheckError(const std::string& str_) : str(new std::string(str_)) {}
+  ~LogCheckError() { if (str != nullptr) delete str; }
+  operator bool() {return str != nullptr; }
+  std::string* str;
+};
+
+
+#define DEFINE_CHECK_FUNC(name, op)                               \
+  template <typename X, typename Y>                               \
+  inline LogCheckError LogCheck##name(const X& x, const Y& y) {   \
+    if (x op y) return LogCheckError();                           \
+    std::ostringstream os;                                        \
+    os << " (" << x << " vs. " << y << ") ";                      \
+    return LogCheckError(os.str());                               \
+  }                                                               \
+  inline LogCheckError LogCheck##name(int x, int y) {             \
+    return LogCheck##name<int, int>(x, y);                        \
+  }
+
+#define CHECK_BINARY_OP(name, op, x, y)                                              \
+  if (scream::util::LogCheckError _check_err = scream::util::LogCheck##name(x, y))   \
+      scream::util::LogMessageFatal(__FILE__, __LINE__).stream()                     \
+      << "Check failed: " << #x " " #op " " #y << *(_check_err.str)
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wsign-compare"
+DEFINE_CHECK_FUNC(_LT, <)
+DEFINE_CHECK_FUNC(_GT, >)
+DEFINE_CHECK_FUNC(_LE, <=)
+DEFINE_CHECK_FUNC(_GE, >=)
+DEFINE_CHECK_FUNC(_EQ, ==)
+DEFINE_CHECK_FUNC(_NE, !=)
+#pragma GCC diagnostic pop
+
+// Always-on checking
+#define SCREAM_CHECK(x)                                           \
+  if (!(x))                                                       \
+    stream::util::LogMessageFatal(__FILE__, __LINE__).stream()    \
+      << "Check failed: " #x << ' '
+#define SCREAM_CHECK_LT(x, y) CHECK_BINARY_OP(_LT, <, x, y)
+#define SCREAM_CHECK_GT(x, y) CHECK_BINARY_OP(_GT, >, x, y)
+#define SCREAM_CHECK_LE(x, y) CHECK_BINARY_OP(_LE, <=, x, y)
+#define SCREAM_CHECK_GE(x, y) CHECK_BINARY_OP(_GE, >=, x, y)
+#define SCREAM_CHECK_EQ(x, y) CHECK_BINARY_OP(_EQ, ==, x, y)
+#define SCREAM_CHECK_NE(x, y) CHECK_BINARY_OP(_NE, !=, x, y)
+#define SCREAM_CHECK_NOTNULL(x) \
+  ((x) == NULL ? scream::util::LogMessageFatal(__FILE__, __LINE__).stream() << "Check  notnull: "  #x << ' ', (x) : (x)) // NOLINT(*)
+
+// Debug-only checking.
+#ifdef NDEBUG
+#define SCREAM_DCHECK(x) \
```

Comment:
You can write simply `#define SCREAM_DCHECK(x)`, ie, with an empty definition, in the case of NDEBUG.